### PR TITLE
[WebGPU] Implement GPUQuerySet

### PR DESCRIPTION
**Pull request by mwyrzykowski**
#### 9ced63bbda6ac9231451b1ca549a16f397e28c78
<pre>
[WebGPU] Implement GPUQuerySet
<a href="https://bugs.webkit.org/show_bug.cgi?id=249356">https://bugs.webkit.org/show_bug.cgi?id=249356</a>
&lt;radar://103379862&gt;

Reviewed by Dean Jackson.

Implement visibility and timestamp queries.

* Source/WebCore/Modules/WebGPU/GPUBuffer.h:
(WebCore::GPUBuffer::~GPUBuffer):
Call the backing&apos;s destroy function so buffer
instances don&apos;t accumulate.

* Source/WebGPU/WebGPU/Buffer.mm:
(WebGPU::Buffer::unmap):
Reset the mapped state on unmap.

* Source/WebGPU/WebGPU/CommandEncoder.h:
* Source/WebGPU/WebGPU/CommandEncoder.mm:
(WebGPU::CommandEncoder::validateRenderPassDescriptor const):
(WebGPU::CommandEncoder::beginRenderPass):
(WebGPU::CommandEncoder::resolveQuerySet):
Resolve visibility and occlusion queries as needed.

(WebGPU::CommandEncoder::writeTimestamp):
writeTimestamp is not practical to implement on TBDR, the
best we can get is granularity of the render pass.

* Source/WebGPU/WebGPU/QuerySet.h:
(WebGPU::QuerySet::create):
(WebGPU::QuerySet::isValid const):
(WebGPU::QuerySet::queryCount const):
(WebGPU::QuerySet::visibilityBuffer const):
(WebGPU::QuerySet::counterSampleBuffer const):
* Source/WebGPU/WebGPU/QuerySet.mm:
(WebGPU::getCounterSet):
(WebGPU::Device::createQuerySet):
(WebGPU::QuerySet::QuerySet):
(WebGPU::QuerySet::destroy):
(WebGPU::QuerySet::setLabel):
(WebGPU::QuerySet::resolveTimestamps const):
Implement timestamp and visibility queries.

* Source/WebGPU/WebGPU/RenderPassEncoder.h:
(WebGPU::RenderPassEncoder::create):
Use timestamps queries if they are present in the descriptor.

* Source/WebGPU/WebGPU/RenderPassEncoder.mm:
(WebGPU::RenderPassEncoder::RenderPassEncoder):
(WebGPU::RenderPassEncoder::beginOcclusionQuery):
(WebGPU::RenderPassEncoder::endOcclusionQuery):
Metal calls for visibility queries.

* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteBuffer.messages.in:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteCommandEncoder.cpp:
(WebKit::RemoteCommandEncoder::copyBufferToBuffer):
Fix typo where source was being used as the source and destination.

* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteQueue.messages.in:
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteBufferProxy.cpp:
(WebKit::WebGPU::RemoteBufferProxy::getMappedRange):
Return early if the buffer was already mapped by mapAsync.

* Websites/webkit.org/demos/webgpu/query-sets.html: Added.
* Websites/webkit.org/demos/webgpu/scripts/query-sets.js: Added.
(async helloTriangle.async frameUpdate):
(async helloTriangle):
Add a demo test for this work.

Canonical link: <a href="https://commits.webkit.org/258236@main">https://commits.webkit.org/258236@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/4f5f31c6629cb580814cc3bd1fcf3c9c3cd84175

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/101310 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/10468 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/34367 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/110583 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/170860 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/105292 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/11411 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/1322 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/93701 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/108411 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/107093 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/8670 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/91910 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/35194 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/90572 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/23316 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/78193 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/4091 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/24830 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/4139 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/1255 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/10239 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/44313 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5668 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/5902 "Built successfully") | | | 
| | | | |

### CODE REVIEW

File: Websites/webkit.org/demos/webgpu/scripts/instanced-textured-cube.js
Comment (mwyrzykowski) [inline Websites/webkit.org/demos/webgpu/scripts/instanced-textured-cube.js:14]: these changes should be reverted

File: Websites/webkit.org/demos/webgpu/scripts/instanced-textured-cube.js
Comment (mwyrzykowski) [inline Websites/webkit.org/demos/webgpu/scripts/instanced-textured-cube.js:255]: these changes should be reverted

File: Websites/webkit.org/demos/webgpu/scripts/instanced-textured-cube.js
Comment (mwyrzykowski) [inline Websites/webkit.org/demos/webgpu/scripts/instanced-textured-cube.js:262]: these changes should be reverted

File: Websites/webkit.org/demos/webgpu/scripts/instanced-textured-cube.js
Comment (mwyrzykowski) [inline Websites/webkit.org/demos/webgpu/scripts/instanced-textured-cube.js:318]: these changes should be reverted

**Comment by mwyrzykowski** [comment]
This is blocked waiting for https://github.com/WebKit/WebKit/pull/7762 to be approved and merged

**Comment by mwyrzykowski** [comment]
https://github.com/WebKit/WebKit/pull/7762 is merged so this can be merged now too after approval

**Comment by grorg** [comment]
Typo in the bug name and commit - QuerySet

**Review by grorg — APPROVED** [review]

File: Websites/webkit.org/demos/webgpu/scripts/query-sets.js
Comment (mwyrzykowski) [inline Websites/webkit.org/demos/webgpu/scripts/query-sets.js:215]: should be `await queue.onSubmittedWorkDone()`

**Pull request merged**
